File: server/scripts/printer.js
const { ThermalPrinter, PrinterTypes } = require('node-thermal-printer');
const fs = require('node:fs');
const path = require('path');

// Configuration paths
const PRINTER_CONFIG_PATH = path.join(__dirname, '../printer-config.json');

// Default printer configuration
const DEFAULT_CONFIG = {
    type: 'EPSON',
    interface: '/dev/usb/lp0',
    width: 60,
    timeout: 5000
};

// Load printer configuration
function loadPrinterConfig() {
    try {
        if (fs.existsSync(PRINTER_CONFIG_PATH)) {
            const configData = fs.readFileSync(PRINTER_CONFIG_PATH, 'utf8');
            return { ...DEFAULT_CONFIG, ...JSON.parse(configData) };
        }
    } catch (error) {
        console.warn('Error loading printer config, using defaults:', error.message);
    }
    return DEFAULT_CONFIG;
}

// Save printer configuration
function savePrinterConfig(config) {
    try {
        const configToSave = { ...DEFAULT_CONFIG, ...config };
        fs.writeFileSync(PRINTER_CONFIG_PATH, JSON.stringify(configToSave, null, 2));
        return true;
    } catch (error) {
        console.error('Error saving printer config:', error);
        return false;
    }
}

// Create printer instance
function createPrinter(config = null) {
    const printerConfig = config || loadPrinterConfig();
    
    return new ThermalPrinter({
        type: PrinterTypes[printerConfig.type] || PrinterTypes.EPSON,
        interface: printerConfig.interface,
        options: {
            timeout: printerConfig.timeout
        },
        width: printerConfig.width
    });
}

// Print raw buffer
async function printBuffer(buffer, config = null) {
    return new Promise((resolve, reject) => {
        let printer;
        try {
            printer = createPrinter(config);
        } catch (error) {
            reject(new Error(`Failed to create printer: ${error.message}`));
            return;
        }

        printer.isPrinterConnected()
            .then(isConnected => {
                if (!isConnected) {
                    throw new Error('Printer is not connected');
                }
                return printer.raw(buffer);
            })
            .then(() => {
                resolve({ success: true, message: 'Print job completed successfully' });
            })
            .catch(error => {
                reject(new Error(`Print failed: ${error.message}`));
            });
    });
}

// Print text with formatting options
async function printText(text, options = {}) {
    return new Promise((resolve, reject) => {
        let printer;
        try {
            printer = createPrinter();
        } catch (error) {
            reject(new Error(`Failed to create printer: ${error.message}`));
            return;
        }

        printer.isPrinterConnected()
            .then(isConnected => {
                if (!isConnected) {
                    throw new Error('Printer is not connected');
                }

                // Apply formatting options
                if (options.align === 'center') printer.alignCenter();
                else if (options.align === 'right') printer.alignRight();
                else printer.alignLeft();

                if (options.bold) printer.bold(true);
                if (options.underline) printer.underline(true);
                if (options.size === 'large') printer.setTextSize(1, 1);

                // Print text
                if (options.newLine !== false) {
                    printer.println(text);
                } else {
                    printer.print(text);
                }

                // Reset formatting
                printer.bold(false);
                printer.underline(false);
                printer.setTextNormal();
                printer.alignLeft();

                if (options.cut) printer.cut();

                return printer.execute();
            })
            .then(() => {
                resolve({ success: true, message: 'Text printed successfully' });
            })
            .catch(error => {
                reject(new Error(`Print failed: ${error.message}`));
            });
    });
}

// Print QR code
async function printQR(data, options = {}) {
    return new Promise((resolve, reject) => {
        let printer;
        try {
            printer = createPrinter();
        } catch (error) {
            reject(new Error(`Failed to create printer: ${error.message}`));
            return;
        }

        printer.isPrinterConnected()
            .then(isConnected => {
                if (!isConnected) {
                    throw new Error('Printer is not connected');
                }

                if (options.align === 'center') printer.alignCenter();
                else if (options.align === 'right') printer.alignRight();

                printer.printQR(data, {
                    cellSize: options.cellSize || 3,
                    correction: options.correction || 'M',
                    model: options.model || 2
                });

                printer.alignLeft();
                if (options.cut) printer.cut();

                return printer.execute();
            })
            .then(() => {
                resolve({ success: true, message: 'QR code printed successfully' });
            })
            .catch(error => {
                reject(new Error(`Print failed: ${error.message}`));
            });
    });
}

// Check printer status
async function getPrinterStatus() {
    try {
        const printer = createPrinter();
        const isConnected = await printer.isPrinterConnected();
        const config = loadPrinterConfig();
        
        return {
            connected: isConnected,
            config: config,
            interface: config.interface
        };
    } catch (error) {
        return {
            connected: false,
            error: error.message,
            config: loadPrinterConfig()
        };
    }
}

module.exports = {
    createPrinter,
    printBuffer,
    printText,
    printQR,
    getPrinterStatus,
    loadPrinterConfig,
    savePrinterConfig
};

File: thermal_printer_api/printer.py
"""Thermal printer management module."""

import base64
from io import BytesIO
from typing import Any, Dict, Optional

import qrcode
import requests
from escpos.printer import Dummy, Network, Serial, Usb
from PIL import Image

from .config import PrinterConfig, load_printer_config


class PrinterManager:
    """Manages thermal printer operations."""

    def __init__(self, config: Optional[PrinterConfig] = None):
        """Initialize printer manager with configuration."""
        self.config = config or load_printer_config()
        self._printer = None

    def _create_printer(self):
        """Create printer instance based on configuration."""
        interface = self.config.interface

        if interface.startswith("/dev/usb/"):
            return Usb(
                idVendor=self.config.idVendor,
                idProduct=self.config.idProduct,
                timeout=0,
                in_ep=self.config.in_ep,
                out_ep=self.config.out_ep,
            )
        elif interface.startswith("/dev/tty"):
            # Serial printer
            return Serial(interface, baudrate=9600, timeout=self.config.timeout / 1000)
        elif interface.startswith("tcp://"):
            # Network printer
            parts = interface.replace("tcp://", "").split(":")
            host = parts[0]
            port = int(parts[1]) if len(parts) > 1 else 9100
            return Network(host=host, port=port, timeout=self.config.timeout / 1000)
        else:
            # Default to dummy printer for testing
            return Dummy()

    def _get_printer(self):
        """Get or create printer instance."""
        if self._printer is None:
            self._printer = self._create_printer()
        return self._printer

    async def check_connection(self) -> bool:
        """Check if printer is connected and accessible."""
        try:
            printer = self._get_printer()
            # For dummy printer, always return True for testing
            if isinstance(printer, Dummy):
                return True
            # For real printers, attempt a simple operation
            return True
        except Exception:
            return False

    async def get_status(self) -> Dict[str, Any]:
        """Get printer status and configuration."""
        try:
            connected = await self.check_connection()
            return {
                "connected": connected,
                "config": self.config.model_dump(),
                "interface": self.config.interface,
            }
        except Exception as e:
            return {
                "connected": False,
                "error": str(e),
                "config": self.config.model_dump(),
            }

    async def print_text(
        self,
        text: str,
        align: str = "left",
        bold: bool = False,
        underline: bool = False,
        size: str = "normal",
        cut: bool = True,
        new_line: bool = True,
    ) -> Dict[str, Any]:
        """Print formatted text."""
        try:
            printer = self._get_printer()

            # Set alignment
            if align == "center":
                printer.set(align="center")
            elif align == "right":
                printer.set(align="right")
            else:
                printer.set(align="left")

            # Set text formatting
            format_args = {}
            if bold:
                format_args["double_height"] = True
                format_args["double_width"] = True
            if underline:
                format_args["underline"] = 1
            if size == "large":
                format_args["double_height"] = True
                format_args["double_width"] = True

            # Apply formatting
            if format_args:
                printer.set(**format_args)

            # Print text
            if new_line:
                printer.text(text + "\n")
            else:
                printer.text(text)

            # Reset formatting
            printer.set()

            if cut:
                printer.cut()

            return {"success": True, "message": "Text printed successfully"}

        except Exception as e:
            return {"success": False, "error": str(e)}

    async def print_qr(
        self,
        data: str,
        align: str = "left",
        cell_size: int = 8,
        correction: str = "M",
        model: int = 2,
        cut: bool = True,
    ) -> Dict[str, Any]:
        """Print QR code."""
        try:
            printer = self._get_printer()

            # Set alignment
            if align == "center":
                printer.set(align="center")
            elif align == "right":
                printer.set(align="right")
            else:
                printer.set(align="left")

            # Create QR code
            qr = qrcode.QRCode(
                version=1,
                error_correction=getattr(
                    qrcode.constants, f"ERROR_CORRECT_{correction}"
                ),
                box_size=cell_size,
                border=4,
            )
            qr.add_data(data)
            qr.make(fit=True)

            # Generate QR code image
            img = qr.make_image(fill_color="black", back_color="white")

            # Convert PIL Image to bytes for escpos
            img_bytes = BytesIO()
            img.save(img_bytes, format="PNG")
            img_bytes.seek(0)

            # Print QR code
            printer.image(img_bytes, impl="bitImageColumn")

            # Reset alignment
            printer.set(align="left")

            if cut:
                printer.cut()

            return {"success": True, "message": "QR code printed successfully"}

        except Exception as e:
            return {"success": False, "error": str(e)}

    async def print_buffer(self, buffer: bytes) -> Dict[str, Any]:
        """Print raw buffer data."""
        try:
            printer = self._get_printer()
            printer._raw(buffer)
            return {"success": True, "message": "Print job completed successfully"}
        except Exception as e:
            return {"success": False, "error": str(e)}

    async def print_image(
        self,
        image_data: str,
        image_type: str,
        align: str = "left",
        impl: str = "bitImageRaster",
        cut: bool = True,
    ) -> Dict[str, Any]:
        """Print image from URL, base64, or file path."""
        try:
            printer = self._get_printer()

            # Set alignment
            if align == "center":
                printer.set(align="center")
            elif align == "right":
                printer.set(align="right")
            else:
                printer.set(align="left")

            # Load image based on type
            img = None
            if image_type == "url":
                response = requests.get(image_data, timeout=10)
                response.raise_for_status()
                img = Image.open(BytesIO(response.content))
            elif image_type == "base64":
                # Remove data URL prefix if present
                if image_data.startswith("data:"):
                    image_data = image_data.split(",", 1)[1]
                img_data = base64.b64decode(image_data)
                img = Image.open(BytesIO(img_data))
            elif image_type == "file":
                img = Image.open(image_data)
            else:
                return {"success": False, "error": "Invalid image_type"}

            # Convert image to bytes for escpos
            img_bytes = BytesIO()
            img.save(img_bytes, format="PNG")
            img_bytes.seek(0)

            # Print image
            printer.image(img_bytes, impl=impl)

            # Reset alignment
            printer.set(align="left")

            if cut:
                printer.cut()

            return {"success": True, "message": "Image printed successfully"}

        except requests.RequestException as e:
            return {
                "success": False,
                "error": f"Failed to fetch image from URL: {str(e)}",
            }
        except Exception as e:
            return {"success": False, "error": str(e)}
